Further insta (#951)
obi1kenobi/cargo-semver-checks: CONTRIBUTING.md

To use a specific toolchain, like beta or nightly, pass it as
an argument: `./scripts/regenerate_test_rustdocs.sh +nightly`.

## What are those `.snap` or `.snap.new` files generated via `cargo test` 

As part of our overall testing strategy, we use a technique called "snapshot testing."
The tool we use for this ([`insta`](https://insta.rs/docs/)) is user friendly and has mutliple ways to interact with it:

These snapshots are by default written to `.snap.new` files (because `INSTA_UPDATE` explained below defaults to `auto`) if they differ and fail the testcase.
Reviewing them is possible via these options:

1. **With `cargo-insta`**: If you install (or have already installed) the `insta` CLI with
   `cargo install --locked cargo-insta`, you can run `cargo insta review`. Check that the
   new output is what you expect, and accept it in the TUI.
2. **Without `cargo-insta`**:

>    From [`insa`s docs](https://insta.rs/docs/quickstart/#tests-without-insta):

Reviewer: ```suggestion
   From [`insta`s docs](https://insta.rs/docs/quickstart/#tests-without-insta):
```

   > You can also just use insta directly from cargo test and control it via the `INSTA_UPDATE` environment variable.
   > The default is auto which will write all new snapshots into .snap.new files if no CI is detected so that cargo-insta can pick them up. The following other modes are possible:
   > - `auto`: the default. no for CI environments or new otherwise
   > - `always`: overwrites old snapshot files with new ones unasked
   > - `unseen`: behaves like always for new snapshots and new for others
   > - `new`: write new snapshots into .snap.new files
   > - `no`: does not update snapshot files at all (just runs tests)

   Thus, if you run the following command, you can accept the current snapshots after reviewing the `.snap.new` files.
   ```text
   INSTA_UPDATE=always cargo test
   ```
3. **Manually**: If you can't (or don't want to) use `cargo-insta`, you can verify the snapshot

>    file manually.  There should be a file called `test_outputs/witnesses/<lint_name>.snap.new`.
>    Open it, and verify that the witnesses generated as expected.  Once you've checked it, move it
>    to `test_outputs/witnesses/<lint_name>.snap` (remove the `.new`)

Reviewer: We don't just use snapshots for witnesses, so let's generalize the advice to anything contributors may run into.
```suggestion
   file manually.  There should be a file called `test_outputs/<some_path>/<lint_name>.snap.new`.
   Open it, and verify that its contents match what you expected: all expected data is present, and no unexpected data is included. 
   Once you've checked it, remove the `.new` suffix so that the file's new path
   is `test_outputs/<some_path>/<lint_name>.snap`
```

To use a specific toolchain, like beta or nightly, pass it as
an argument: `./scripts/regenerate_test_rustdocs.sh +nightly`.

> ## What are those `.snap` or `.snap.new` files generated via `cargo test` 

Reviewer: Does it make sense to make CI fail if `.snap.new` files are checked into the repo? Will it do that already?

Author: It will somewhat already do that (unless one creates a `.snap.new` and the same `.snap` file, will have to test that part)

Reviewer: Yes, having both a `.snap.new` and a `.snap` file was what I would love to make sure we prevent. Because if that happens, it will pollute the repo and confuse the local setups of everyone else, since everyone's insta may think there are altered snapshots.

> Next to the regular testcase strategies, we also utilise snapshot testing.

Reviewer: ```suggestion
As part of our overall testing strategy, we use a technique called "snapshot testing."
```

   > You can also just use insta directly from cargo test and control it via the `INSTA_UPDATE` environment variable.
   > The default is auto which will write all new snapshots into .snap.new files if no CI is detected so that cargo-insta can pick them up. The following other modes are possible:
   > - `auto`: the default. no for CI environments or new otherwise
   > - `always`: overwrites old snapshot files with new ones unasked
   > - `unseen`: behaves like always for new snapshots and new for others
   > - `new`: write new snapshots into .snap.new files
   > - `no`: does not update snapshot files at all (just runs tests)

   Thus, if you run the following command, you can accept the current snapshots after reviewing the `.snap.new` files.
   ```text
   INSTA_UPDATE=always cargo test
   ```
3. **Manually**: If you can't (or don't want to) use `cargo-insta`, you can verify the snapshot
   file manually.  There should be a file called `test_outputs/witnesses/<lint_name>.snap.new`.
   Open it, and verify that the witnesses generated as expected.  Once you've checked it, move it
   to `test_outputs/witnesses/<lint_name>.snap` (remove the `.new`)

> 
> 

Reviewer: Please try to review your own PRs using the GitHub interface before marking them ready for review, so that you can catch and fix little issues like trailing whitespace or other things that become obvious in the diff viewer. It'll save time and will mean we can merge your PRs faster.
```suggestion
```

         31 │+      "path": List([
         32 │+        String("enum_struct_variant_field_added"),
         33 │+        String("PubEnum"),
         34 │+      ]),
         35 │+      "span_begin_line": Uint64(4),
         36 │+      "span_filename": String("src/lib.rs"),
         37 │+      "variant_name": String("Foo"),
         38 │+    },
         39 │+  ],
    2    40 │ }
────────────┴───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
To update snapshots run `cargo insta review`
Stopped on the first failure. Run `cargo insta test` to run all snapshots.
thread 'query::tests_lints::enum_struct_variant_field_added' panicked at /home/frank/.cargo/registry/src/index.crates.io-6f17d22bba15001f/insta-1.40.0/src/runtime.rs:548:9:
```

Inspect the generated "actual" output in the `.snap.new` file:
- Does it report the semver issue your lint was supposed to catch? If not, the lint query
  or the test crates' code may need to be tweaked.
- Does it report correct span information? Is the span as specific as possible, for example

>   pointing to a `struct`'s field rather than the whole struct if the lint refers to that field?

Author: yea, sorry thought of this as a typo-fix

- Does the output also report any code from test crates other than `test_crates/<lint_name>`?
  If so, ensure the reported code is indeed violating semver and is not being flagged
  by any other lint.

> If everything looks okay, either run `cargo insta review` (see the [snapshot instructions](#what-are-those-snap-or-snapnew-files-generated-via-cargo-test-) for context) or manually move `test_outputs/query_execution/<lint_name>.snap.new` to  `test_outputs/query_execution/<lint_name>.snap`.

Author: Interesting, seems like github and rustrover handle this part of markdown differently
(github removes trailing dashes, rustrover wants them)

![image](https://github.com/user-attachments/assets/01e0bb94-cab7-4815-a853-d9b626deccf7)
![image](https://github.com/user-attachments/assets/e6b716b7-b6ad-43b1-bf89-4583a2d0eb4f)

it will fail, because there's no expected result to compare to.  Let's make the test pass:

We use `insta` for snapshot testing witness results, so after adding/changing a witness, we need
to update the test outputs.

> [!TIP]

> > It may contain output for other test crates - this is not necessarily an error:

Reviewer: this is just me nitpicking on typography, don't mind me 😅
```suggestion
> It may contain output for other test crates — this is not necessarily an error:
```

it will fail, because there's no expected result to compare to.  Let's make the test pass:

We use `insta` for snapshot testing witness results, so after adding/changing a witness, we need
to update the test outputs.

> [!TIP]
> It may contain output for other test crates - this is not necessarily an error:

> > See the [troubleshooting section](#troubleshooting) for more info.

Reviewer: Nice callout!
